compiler/hash-ir/src/ir.rs

```diff
@@ -155,33 +160,77 @@ pub enum AddressMode {
     Smart,
 }
 
-/// A defined statement within the IR
 #[derive(Debug, PartialEq, Eq)]
-pub enum StatementKind<'i> {
-    /// Filler kind when expressions are optimised out or removed for other
-    /// reasons.
-    Nop,
+pub enum PlaceProjection {
+    /// When we want to narrow down the union type to some specific
+    /// variant.
+    Downcast(usize),
+    /// A reference to a specific field within the place, at this stage they
+    /// are represented as indexes into the field store of the place type.
+    Field(usize),
+    /// Take the index of some specific place, the index does not need to be
+    /// constant
+    Index(Local),
+    /// We want to dereference the place
+    Deref,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub struct Place {
+    /// The original place of where this is referring to.
+    pub local: Local,
+    /// Any projections that are applied onto the `local` in
+    /// order to specify an exact location within the local.
+    pub projections: Vec<PlaceProjection>,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum AggregateKind {
+    Tuple,
+    Array(TermId),
+    Enum(NominalDefId, usize),
+    Struct(NominalDefId),
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum RValue<'ir> {
     /// A constant value.
     Const(Const),
 
-    /// A local variable value
-    Local(Local),
+    /// A local variable value, do we need to denote whether this is a
+    /// copy/move?
+    Use(Place),
+
+    /// Compiler intrinsic operation, this will be computed in place and
+    /// replaced by a constant.
+    ///
+    /// @@Future: maybe in the future this should be replaced by a compile-time
+    /// API variant which will just run some kind of operation and return the
+    /// constant.
+    ConstOp(ConstOp, &'ir RValue<'ir>),
```

Comment:
How does sizeof work? How does it know the type of the value

Reply:
There is an associated term with the value, so it can always compute the size of the term, that is if the term is "sized", otherwise it should panic since we shouldn't be getting un-sized type arguments here

Reply:
But even still, given any RValue, how can one procedurally figure out its type? (And thus its size)

Reply:
It should probably just take a `TermId`

Reply:
Hmm, is it a good idea to let TC concepts leak into here? Isn't there a way of having a more generalised way to represent type layouts (separate crate) and use that in here rather than relying on all the TC machinery?

Reply:
Well, it would need to convert to some other representation of types before it reaches this stage. In my opinion, this is quite the task since it would mean that we would have to represent the whole type system in a different representation than the current one. 

I think this work could be done in the future (where this system is as simple as possible, and erases most un-needed things, possibly even smaller in terms of memory footprint), but I don't think this is the current priority.

Reply:
The only problem is that having TC types in the IR means that the IR needs to be able to handle all of the representable types (or alternatively assert that only a subset of the types exist at this stage). Surely not the entire type system, but rather a small part of it, needs to be representable in the IR.